### ISSUE TITLE
Watcher on configuration file

### ISSUE THREAD

**Issue by nuest** [52North/iceland#15]
Where could I integrate a "watcher" mechanism that regularly (e.g. every 10 seconds) checks the configuration file for changes? @autermann 

In my scenario, I do not have a user interface for changing settings, but I still want an admin to be able to change setting values without having to restart the service.

**Comment by autermann** [comment]
Mmh... the `JSONConfiguration` and DAOs can not push changes up to the `SettingsService`, so you have to implement this in the `SettingsService`. Just subclass it and replace the bean with your own implementation. Then extend the DAO classes and the `JSONConfiguration` with some reload mechanism and replace their beans...

**Comment by nuest** [comment]
@autermann This will make things easier for me so I'd like to implement it. The problem I see is that ``JsonSettingsDao`` knows the file (and can thereby [watch it](http://andreinc.net/2013/12/06/java-7-nio-2-tutorial-writing-a-simple-filefolder-monitor-using-the-watch-service-api/)), but ``SettingsServiceImpl`` is the class that "propagates" the settings. We don't want a circular dependency here, right?

Should I introduce a new component that knows both the SettingsService and the SettingsDao, which is notified by the latter if the settings change and then causes the service to configure all objects again?

**Comment by nuest** [comment]
Or should we use the service event bus for this?

**Comment by nuest** [comment]
@autermann I lost my notes on this but I hope I remember the plan correctly: 

* Implement a new third component that watches the configuration file for changes
  * ``WatchingJsonConfiguration extends JsonConfiguration``, it has the additional constructor parameter ``watchInterval``
  * it must extend ``JsonConfiguration`` because that is the only place where the filename is known
  * alternatively it could be a third component also getting injected the configlocationprovider, but then the file name would have to be kept in sync in ``application-context.xml`` manually. BUT this third component could know both the JsonSettingsDao and the ``SettingsService`` and thereby, when the file is changed, re-apply all settings.
  * probably has to be the third component which cannot extend ``JsonConfiguration``. In that case, is there a method to call ``SettingsService.applySetting()`` for all existing settings? Or to reconfigure all objects? It should imho be in ``SettingsService``...
  * when the file is changed, the ``refresh()`` method is called
* do not use the event bus (though looking at the code I don't remember why not...)
* somehow ``SettingsService`` needs to re-configure all objects

**Comment by autermann** [comment]
A third component ist the way to go, as a `WatchingJSONConfiguration` would probably introduce a circular dependency to the `SettingsManager`. You could circumvent this by using the `EventBus`, but I think the `EventBus` would be the wrong place for something like this. IMHO it is more like a extension API to hook into the internal processes, it just sounds wrong to me :)

Regarding the file name: just introduce a getter for `file` or `fileName` in `JsonConfiguration`.

Your component watches the file for modifications. If a modification happes, it calls `JsonConfiguration#refresh()`. You could then call `SettingsService#changeSetting()` for each setting that `JSONSettingsDao#getSettingValues()` returns, but that will not detect deleted settings and will probably be quite a mess, because every call will eventually persist the setting using `JSONSettingsDao` :-1: 

So you should introduce a new method (like `SettingsService#reconfigure()`) that iterates over all `ConfigurableObject`s and reapplies the settings (at this point you won't be able to check if a settings was changed, so it will only be something likes this (plus exception handling ;))

```java
@Override
public void reconfigure() {
	this.configurableObjectsLock.readLock().lock();
	try {
	    configurableObjects.values().stream()
			.forEach(co -> co.configure(getNotNullSettingValue(co));)
	} finally {
	    this.configurableObjectsLock.readLock().unlock();
	}
}
```

**Comment by autermann** [comment]
Or more efficient, refactor `getNotNullSettingValue()` to accept `String` keys and use something like this:

```java
@Override
public void reconfigure() {
	this.configurableObjectsLock.readLock().lock();
	try {
		configurableObjects.keySet().stream()
			.map(this::getNotNullSettingValue)
			.forEach(sv -> configurableObjects.get(sv.getKey()).stream()
								.forEach(co -> co.configure(sv)));
	} finally {
	    this.configurableObjectsLock.readLock().unlock();
	}
}
```

**Comment by autermann** [comment]
Fixed in #30.
